Backport fix for CVE-2026-108359 (unbounded ODS row and column repeats) from 5.10.0

diff --git a/src/PhpSpreadsheet/Reader/Ods.php b/src/PhpSpreadsheet/Reader/Ods.php
index 4ddbae0..f064d21 100644
--- a/src/PhpSpreadsheet/Reader/Ods.php
+++ b/src/PhpSpreadsheet/Reader/Ods.php
@@ -209,6 +209,7 @@ class Ods extends BaseReader
                     if ($xml->name == 'table:table-row' && $xml->nodeType == XMLReader::ELEMENT) {
                         $rowspan = $xml->getAttribute('table:number-rows-repeated');
                         $rowspan = empty($rowspan) ? 1 : (int) $rowspan;
+                        self::checkRowsRepeated($currRow, $rowspan);
                         $currRow += $rowspan;
                         $currCol = 0;
                         // Step into the row
@@ -218,6 +219,10 @@ class Ods extends BaseReader
                             if ($xml->name == 'table:table-cell' && $xml->nodeType == XMLReader::ELEMENT) {
                                 $mergeSize = $xml->getAttribute('table:number-columns-repeated');
                                 $mergeSize = empty($mergeSize) ? 1 : (int) $mergeSize;
+                                self::checkColumnsRepeatedInt(
+                                    $currCol,
+                                    $mergeSize
+                                );
                                 $currCol += $mergeSize;
                                 if (!$xml->isEmptyElement) {
                                     $tmpInfo['totalColumns'] = max($tmpInfo['totalColumns'], $currCol);
@@ -227,6 +232,11 @@ class Ods extends BaseReader
                                 }
                             } elseif ($xml->name == 'table:covered-table-cell' && $xml->nodeType == XMLReader::ELEMENT) {
                                 $mergeSize = $xml->getAttribute('table:number-columns-repeated');
+                                $mergeSize = empty($mergeSize) ? 1 : (int) $mergeSize;
+                                self::checkColumnsRepeatedInt(
+                                    $currCol,
+                                    $mergeSize
+                                );
                                 $currCol += (int) $mergeSize;
                             }
                             if ($doread) {
@@ -322,7 +332,7 @@ class Ods extends BaseReader
                 ->scan($zip->getFromName('meta.xml'))
         );
         if ($xml === false) {
-            throw new Exception('Unable to read data from {$pFilename}');
+            throw new Exception("Unable to read data from {$filename}");
         }
 
         /** @var array{meta?: string, office?: string, dc?: string} */
@@ -817,6 +827,7 @@ class Ods extends BaseReader
         } else {
             $rowRepeats = 1;
         }
+        self::checkRowsRepeated($rowID, $rowRepeats);
         $worksheet = $spreadsheet->getSheetByName($worksheetName);
 
         $columnID = 'A';
@@ -830,6 +841,8 @@ class Ods extends BaseReader
             } else {
                 $colRepeats = 1;
             }
+            $columnIndex = Coordinate::columnIndexFromString($columnID);
+            self::checkColumnsRepeated($columnID, $colRepeats);
             $styleName = $cellData->getAttributeNS($tableNs, 'style-name');
             if ($styleName === '') {
                 if ($worksheet === null || !$worksheet->columnDimensionExists($columnID)) {
@@ -1313,17 +1326,19 @@ class Ods extends BaseReader
         bool $processStyles = true
     ): void {
         if ($childNode->hasAttributeNS($tableNs, 'number-columns-repeated')) {
-            $rowRepeats = (int) $childNode->getAttributeNS($tableNs, 'number-columns-repeated');
+            $colRepeats = (int) $childNode->getAttributeNS($tableNs, 'number-columns-repeated');
         } else {
-            $rowRepeats = 1;
+            $colRepeats = 1;
         }
+        // called routine expects index to be 1 less than it is
+        self::checkColumnsRepeatedInt($tableColumnIndex - 1, $colRepeats);
         $tableStyleName = $childNode->getAttributeNS($tableNs, 'style-name');
         if ($processWidths) {
             if (isset($columnWidths[$tableStyleName])) {
                 $columnWidth = new HelperDimension($columnWidths[$tableStyleName]);
                 $tableColumnIndex2 = $tableColumnIndex;
                 $tableColumnString = Coordinate::stringFromColumnIndex($tableColumnIndex2);
-                for ($rowRepeats2 = $rowRepeats; $rowRepeats2 > 0 && $tableColumnIndex2 <= AddressRange::MAX_COLUMN_INT; --$rowRepeats2) {
+                for ($colRepeats2 = $colRepeats; $colRepeats2 > 0 && $tableColumnIndex2 <= AddressRange::MAX_COLUMN_INT; --$colRepeats2) {
                     if (!$this->readEmptyCells && $tableColumnIndex2 > $this->highestDataIndex) {
                         break;
                     }
@@ -1342,7 +1357,7 @@ class Ods extends BaseReader
             if ($defaultStyleName !== 'Default' && isset($this->allStyles[$defaultStyleName])) {
                 $tableColumnIndex2 = $tableColumnIndex;
                 $tableColumnString = Coordinate::stringFromColumnIndex($tableColumnIndex2);
-                for ($rowRepeats2 = $rowRepeats; $rowRepeats2 > 0 && $tableColumnIndex2 <= AddressRange::MAX_COLUMN_INT; --$rowRepeats2) {
+                for ($colRepeats2 = $colRepeats; $colRepeats2 > 0 && $tableColumnIndex2 <= AddressRange::MAX_COLUMN_INT; --$colRepeats2) {
                     $spreadsheet->getActiveSheet()
                         ->getStyle($tableColumnString)
                         ->applyFromArray(
@@ -1355,7 +1370,7 @@ class Ods extends BaseReader
                 }
             }
         }
-        $tableColumnIndex += $rowRepeats;
+        $tableColumnIndex += $colRepeats;
     }
 
     private function processSettings(ZipArchive $zip, Spreadsheet $spreadsheet): void
@@ -1801,4 +1816,28 @@ class Ods extends BaseReader
             }
         }
     }
+
+    private static function checkRowsRepeated(int $rowID, int $rowRepeats): void
+    {
+        if ($rowRepeats < 1 || $rowID + $rowRepeats - 1 > AddressRange::MAX_ROW) {
+            throw new Exception("Invalid number-rows-repeated $rowRepeats following row $rowID");
+        }
+    }
+
+    private static function checkColumnsRepeated(string $colID, int $colRepeats): void
+    {
+        $colIndex = Coordinate::columnIndexFromString($colID);
+        if ($colRepeats < 1 || $colIndex + $colRepeats - 1 > AddressRange::MAX_COLUMN_INT) {
+            throw new Exception("Invalid number-columns-repeated $colRepeats following column $colID");
+        }
+    }
+
+    private static function checkColumnsRepeatedInt(int $colIndex, int $colRepeats): void
+    {
+        // We don't have column string at this point,
+        //    and colIndex is actually 1 less than it should be.
+        if ($colRepeats < 1 || $colIndex + $colRepeats > AddressRange::MAX_COLUMN_INT) {
+            throw new Exception("Invalid number-columns-repeated $colRepeats following column index $colIndex");
+        }
+    }
 }
